src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIEnumerableDefaultConverter.cs

```diff
@@ -0,0 +1,297 @@
+    /// </summary>
+    internal abstract class JsonIEnumerableDefaultConverter<TCollection, TElement> : JsonIEnumerableConverter<TCollection, TElement>
+    {
+        protected abstract void Add(TElement value, ref ReadStack state);
+
+        protected virtual void CreateCollection(ref ReadStack state, JsonSerializerOptions options) { }
+        protected virtual void ConvertCollection(ref ReadStack state, JsonSerializerOptions options) { }
+
+        protected static JsonConverter<TElement> GetElementConverter(ref ReadStack state)
+        {
+            JsonConverter<TElement>? converter = state.Current.JsonClassInfo.ElementClassInfo!.PolicyProperty!.ConverterBase as JsonConverter<TElement>;
+            if (converter == null)
+            {
+                state.Current.JsonClassInfo.ElementClassInfo.PolicyProperty.ThrowCollectionNotSupportedException();
+            }
+
+            return converter!;
+        }
+
+        protected static JsonConverter<TElement> GetElementConverter(ref WriteStack state)
+        {
+            JsonConverter<TElement>? converter = state.Current.DeclaredJsonPropertyInfo.ConverterBase as JsonConverter<TElement>;
+            if (converter == null)
+            {
+                state.Current.JsonClassInfo.ElementClassInfo!.PolicyProperty!.ThrowCollectionNotSupportedException();
+            }
+
+            return converter!;
+        }
+
+        internal override bool OnTryRead(
+            ref Utf8JsonReader reader,
+            Type typeToConvert,
+            JsonSerializerOptions options,
+            ref ReadStack state,
+            out TCollection value)
+        {
+            bool shouldReadPreservedReferences = options.ReferenceHandling.ShouldReadPreservedReferences();
+
+            if (!state.SupportContinuation && !shouldReadPreservedReferences)
+            {
+                // Fast path that avoids maintaining state variables and dealing with preserved references.
+
+                if (reader.TokenType != JsonTokenType.StartArray)
+                {
+                    ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(TypeToConvert);
+                }
+
+                CreateCollection(ref state, options);
+
+                JsonConverter<TElement> elementConverter = GetElementConverter(ref state);
+                if (elementConverter.CanUseDirectReadOrWrite)
+                {
+                    // Fast path that avoids validation and extra indirection.
+                    while (true)
+                    {
+                        reader.Read();
+                        if (reader.TokenType == JsonTokenType.EndArray)
+                        {
+                            break;
+                        }
+
+                        // Obtain the CLR value from the JSON and apply to the object.
+                        TElement element = elementConverter.Read(ref reader, elementConverter.TypeToConvert, options);
```

Comment:
<details>
<summary>
It turns out we don't pass `null` to converters in any instance. This is generally what we did in 3.0/3.1. There's no action here for this PR, although there's an open question regarding whether we should, at the expense of performance. See the source below for more info on the current behavior (passing null to converters should cause `NIE`, but none occurs).
</summary>
```C#
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

public class Program
{
	public static void Main()
	{
		var options = new JsonSerializerOptions();
		options.Converters.Add(new JsonSimpleClassConverter());
                options.Converters.Add(new JsonStringConverter());
		
		// Converters not for null collection elements
		JsonSerializer.Deserialize<List<SimpleClass>>("[null]", options);
		
		JsonSerializer.Serialize("[null]", options);
		
		// Converters not called for null root objects
		JsonSerializer.Deserialize<SimpleClass>("null");
		
		JsonSerializer.Serialize<SimpleClass>(null);
		
		SimpleClass obj = null;
		JsonSerializer.Serialize(obj, typeof(SimpleClass));
		
		// Converters not called for null properties
		JsonSerializer.Deserialize<ParentOfSimpleClass>(@"{""SimpleClass"":null}");
		
		JsonSerializer.Serialize<ParentOfSimpleClass>(new ParentOfSimpleClass());
		JsonSerializer.Serialize(new ParentOfSimpleClass(), typeof(ParentOfSimpleClass));

                // Converters not for null collection elements
                JsonSerializer.Deserialize<List<string>>("[null]", options);

                string str = null;
                JsonSerializer.Serialize(str, options);

                // Converters not called for null root objects
                JsonSerializer.Deserialize<string>("null");

                JsonSerializer.Serialize<string>(null);

                string obj = null;
                JsonSerializer.Serialize(obj, typeof(string));

                // Converters not called for null properties
                JsonSerializer.Deserialize<ParentOfString>(@"{""String"":null}");

                JsonSerializer.Serialize<ParentOfString>(new ParentOfString());
                JsonSerializer.Serialize(new ParentOfString(), typeof(ParentOfString));
	}
	
	public class ParentOfSimpleClass
	{
		public SimpleClass SimpleClass;
	}
	
	public class SimpleClass { }
	
	public class JsonSimpleClassConverter : JsonConverter<SimpleClass>
	{
		public override SimpleClass Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
		{
			throw new NotImplementedException();
		}
		
		public override void Write(Utf8JsonWriter writer, SimpleClass @value, JsonSerializerOptions options)
		{
			throw new NotImplementedException();
		}
	}

        public class ParentOfString
        {
                public string String;
        }

        public class JsonStringConverter : JsonConverter<string>
        {
                public override string Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
                {
                        throw new NotImplementedException();
                }

                public override void Write(Utf8JsonWriter writer, string @value, JsonSerializerOptions options)
                {
                        throw new NotImplementedException();
                }
        }
}
```
<details>

Reply:
We will pass tokenType.Null to converters that can't be null normally (i.e. structs). The Nullable and JsonElement converters leverage this since they want to treat null differently. Having converters that normally allow `null` (i.e. classes) to be required to process `null` token type would likely cause lots of converter bugs, so the framework handles it, plus handles it more efficiently higher up the stack.

FWIW this PR adds internal `bool HandleNullValue` property to `JsonConverter` so in theory a converter could elect to process null.